news/tests.py

```diff
@@ -214,6 +221,80 @@ def toggle(pk, attr):
         self.assertEquals(toggle(self.article.pk, 'hidden'), {'color': 'grey' if hidden else 'yellow'})
         self.assertEquals(toggle(self.article.pk, 'hidden'), {'color': 'yellow' if hidden else 'grey'})
 
+    def test_event_ticket_emails_only_returns_active_tickets_emails(self):
+        username_and_ticket_state_tuples = [
+            ("user2", True),
+            ("user3", False),
+            ("user4", True),
+        ]
+
+        tickets = create_tickets_for(
+            event=self.event,
+            username_and_ticket_state_tuples=username_and_ticket_state_tuples
+        )
+        expected_ticket_emails = "user2@example.com,user4@example.com"
+        self.add_permission("change_event")
+        
+        response = self.client.get(reverse('event-tickets', args=[self.event.pk]))
+        
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(expected_ticket_emails, response.context["ticket_emails"])
+
+    def test_timeplace_ticket_emails_only_returns_active_tickets_emails(self):
+        username_and_ticket_state_tuples = [
+            ("user2", True),
+            ("user3", False),
+            ("user4", True),
+        ]
+
+        tickets = create_tickets_for(
+            event=self.timeplace,
+            username_and_ticket_state_tuples=username_and_ticket_state_tuples
+        )
+        expected_ticket_emails = "user2@example.com,user4@example.com"
+        self.add_permission("change_event")
+        
+        response = self.client.get(reverse('timeplace-tickets', args=[self.timeplace.pk]))
+        
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(expected_ticket_emails, response.context["ticket_emails"])
```

Comment:
There is a lot of code duplication here. Is there a way of parametrising the test? The only difference between these two tests are swapping out `event` for `timeplace`. I know this is possible with pytest, but it [does not look compatible](https://docs.pytest.org/en/latest/unittest.html#pytest-features-in-unittest-testcase-subclasses) with `unittest.TestCase`

Reply:
I agree, and I guess the easiest thing to do would be to simply extract the common code to a separate method with an `event` and a `url_name` parameter (for the call to `reverse()`), and a name starting with e.g. `_test_...` (as methods not starting with `test` are not run), and calling this method from the two methods you've highlighted.

Also, while we're talking about it, it would be nice if you could move the `create_tickets_for()` function to just below the methods that are using it, indenting it (so that it's part of the `ViewTestCase` class), and annotating it with `@staticmethod` 🙂 It's easier to read the code if the methods calling each other are next to one another.

Reply:
I created a method called `assert_context_ticket_emails` that does most of the work. This method is called by the original tests. Even though it did not start with test, it looks like it runs, because it was called from the original test functions.

Reply:
Nice! 😄 

> Even though it did not start with test, it looks like it runs, because it was called from the original test functions.

When I wrote "run", I meant that the method is not called directly by the test runner; however, any test method calling that function will naturally make it run, so yes, you're right 🙂